server.go

```diff
@@ -601,41 +606,86 @@ type nodeAddresses struct {
 // sendToPeer send a message to the server telling it to send the specific set
 // of message to a particular peer. If the peer connect be found, then this
 // method will return a non-nil error.
-func (s *server) sendToPeer(target *btcec.PublicKey, msgs ...lnwire.Message) error {
-	errChan := make(chan error, 1)
-
+// handleSend attempts to forward a sendReq to the outgoing queue of a
+// particular peer. Access to the peers map is synchronized to prevent changes
+// in set membership during lookup. The process of queueing the message and
+// signaling back to the caller is done asynchronously to free up the
+// queryHandler.
+func (s *server) SendToPeer(target *btcec.PublicKey, msgs ...lnwire.Message) error {
 	msgsToSend := make([]lnwire.Message, 0, len(msgs))
 	msgsToSend = append(msgsToSend, msgs...)
-	sMsg := &sendReq{
-		target:  target,
-		msgs:    msgsToSend,
-		errChan: errChan,
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.sendToPeer(target, msgsToSend)
+}
+
+func (s *server) sendToPeer(target *btcec.PublicKey, msgs []lnwire.Message) error {
+	// Compute the target peer's identifier.
+	targetPubBytes := target.SerializeCompressed()
+
+	srvrLog.Infof("Attempting to send msgs %v to: %x",
+		len(msgs), targetPubBytes)
+
+	// Lookup intended target in peersByPub, returning an error to the caller if
+	// the peer is unknown.  Access to peersByPub is synchronized here to ensure
+	// we consider the exact set of peers present at the time the queryHandler
+	// accepted the broadcastReq.
+	targetPeer, ok := s.peersByPub[string(targetPubBytes)]
+	if !ok {
+		srvrLog.Errorf("unable to send message to %x, "+
+			"peer not found", targetPubBytes)
+
+		return errors.New("peer not found")
 	}
 
-	select {
-	case s.sendRequests <- sMsg:
-	case <-s.quit:
-		return errors.New("server shutting down")
+	s.sendPeerMessages(targetPeer, msgs, nil)
+
+	return nil
+}
+
+func (s *server) sendPeerMessages(
+	targetPeer *peer,
+	msgs []lnwire.Message,
+	wg *sync.WaitGroup) {
+
+	if wg != nil {
+		defer wg.Done()
+		defer s.wg.Done()
 	}
 
-	select {
-	case err := <-errChan:
-		return err
-	case <-s.quit:
-		return errors.New("server shutting down")
+	for _, msg := range msgs {
+		done := make(chan struct{})
+		select {
+		case targetPeer.outgoingQueue <- outgoinMsg{msg, done}:
+			select {
+			case <-done:
```

Comment:
I don't think the process of sending messages to a peer should be fully asynchronous. Currently, this API is by:
  * The `AuthenticatedGossiper` when broadcasting new node announcements
  * The `ChannelRouter` when syncing up a newly connected peer to our graph state
  * The `fundingManager` when processing the workflow of an active funding reservation 

In my mind, none of these cases require synchronous sends.